examples/generative/molecule_generation.py

```diff
@@ -0,0 +1,575 @@
+        return {"loss": self.train_total_loss_tracker.result()}
+
+    def calculate_loss(
+        self, z_log_var, z_mean, qed_true, qed_pred, graph_real, graph_generated
+    ):
+
+        adjacency_real, features_real = graph_real
+        adjacency_gen, features_gen = graph_generated
+
+        adjacency_loss = tf.reduce_mean(
+            tf.reduce_sum(
+                keras.losses.categorical_crossentropy(adjacency_real, adjacency_gen),
+                axis=(1, 2),
+            )
+        )
+        features_loss = tf.reduce_mean(
+            tf.reduce_sum(
+                keras.losses.categorical_crossentropy(features_real, features_gen),
+                axis=(1),
+            )
+        )
+        kl_loss = -0.5 * tf.reduce_sum(
+            1 + z_log_var - tf.square(z_mean) - tf.exp(z_log_var), 1
+        )
+        kl_loss = tf.reduce_mean(kl_loss)
+
+        property_loss = tf.reduce_mean(
+            keras.losses.binary_crossentropy(qed_true, qed_pred)
+        )
+
+        graph_loss = self._gradient_penalty(graph_real, graph_generated)
+
+        return kl_loss + property_loss + graph_loss + adjacency_loss + features_loss
+
+    def _gradient_penalty(self, graph_real, graph_generated):
+        # Unpack graphs
+        adjacency_real, features_real = graph_real
+        adjacency_generated, features_generated = graph_generated
+
+        # Generate interpolated graphs (adjacency_interp and features_interp)
+        alpha = tf.random.uniform([self.batch_size])
+        alpha = tf.reshape(alpha, (self.batch_size, 1, 1, 1))
+        adjacency_interp = (adjacency_real * alpha) + (1 - alpha) * adjacency_generated
+        alpha = tf.reshape(alpha, (self.batch_size, 1, 1))
+        features_interp = (features_real * alpha) + (1 - alpha) * features_generated
+
+        # Compute the logits of interpolated graphs
+        with tf.GradientTape() as tape:
+            tape.watch(adjacency_interp)
+            tape.watch(features_interp)
+            _, _, logits, _, _ = self(
+                [adjacency_interp, features_interp], training=True
+            )
+
+        # Compute the gradients with respect to the interpolated graphs
+        grads = tape.gradient(logits, [adjacency_interp, features_interp])
+        # Compute the gradient penalty
+        grads_adjacency_penalty = (1 - tf.norm(grads[0], axis=1)) ** 2
+        grads_features_penalty = (1 - tf.norm(grads[1], axis=2)) ** 2
+        return tf.reduce_mean(
+            tf.reduce_mean(grads_adjacency_penalty, axis=(-2, -1))
+            + tf.reduce_mean(grads_features_penalty, axis=(-1))
+        )
+
+    def call(self, inputs):
+        z_mean, log_var = self.encoder(inputs)
+        z = Sampling()([z_mean, log_var])
+
+        gen_adjacency, gen_features = self.decoder(z)
+
+        property_pred = self.property_prediction_layer(z_mean)
+
+        return z_mean, log_var, property_pred, gen_adjacency, gen_features
```

Comment:
Should we return a dict here? 5 different outputs is difficult to manage by position only.